src/drafts/SelectPanel2/SelectPanel.tsx

```diff
@@ -0,0 +1,310 @@
+      >
+        {/* TODO: Keyboard navigation of actionlist should be arrow keys
+            with tabs to enter and escape
+        */}
+        <SelectPanelContext.Provider
+          value={{
+            onCancel: onInternalClose,
+            onClearSelection: onInternalClearSelection,
+            searchQuery,
+            // @ts-ignore todo
+            setSearchQuery,
+          }}
+        >
+          <Box as="form" onSubmit={onInternalSubmit} sx={{display: 'flex', flexDirection: 'column', height: '100%'}}>
+            {contents}
+          </Box>
+        </SelectPanelContext.Provider>
+      </AnchoredOverlay>
+    </>
+  )
+}
+
+const SelectPanelButton = React.forwardRef((props, anchorRef) => {
+  // @ts-ignore todo
+  return <Button ref={anchorRef} {...props} />
+})
+SelectPanel.Button = SelectPanelButton
+
+const SelectPanelHeader: React.FC<React.PropsWithChildren> = ({children, ...props}) => {
+  const [slots] = useSlots(children, {
+    heading: SelectPanelHeading,
+    searchInput: SelectPanelSearchInput,
+  })
+
+  const {onCancel, onClearSelection} = React.useContext(SelectPanelContext)
+
+  return (
+    <Box id="header" sx={{padding: 2, borderBottom: '1px solid', borderColor: 'border.default'}} {...props}>
+      <Box sx={{display: 'flex', justifyContent: 'space-between', alignItems: 'center', paddingBottom: 2}}>
+        {slots.heading}
+        <Box>
+          {/* Will not need tooltip after https://github.com/primer/react/issues/2008 */}
+          <Tooltip text="Clear selection" direction="s" onClick={onClearSelection}>
+            <IconButton type="button" variant="invisible" icon={ClearIcon} aria-label="Clear selection" />
+          </Tooltip>
+          <Tooltip text="Close" direction="s">
+            <IconButton type="button" variant="invisible" icon={XIcon} aria-label="Close" onClick={() => onCancel()} />
+          </Tooltip>
+        </Box>
+      </Box>
+      {slots.searchInput}
+    </Box>
+  )
+}
+SelectPanel.Header = SelectPanelHeader
+
+const SelectPanelHeading: React.FC<React.PropsWithChildren<{children: string}>> = ({children, ...props}) => {
+  // heading element is intentionally hardcoded to h1, it is not customisable
+  // see https://github.com/github/primer/issues/2578 for context
+  return (
+    <Heading as="h1" sx={{fontSize: 14, fontWeight: 600, marginLeft: 2}} {...props}>
+      {children}
+    </Heading>
+  )
+}
+SelectPanel.Heading = SelectPanelHeading
+
+// @ts-ignore todo
+const SelectPanelSearchInput = props => {
+  const inputRef = React.createRef<HTMLInputElement>()
+
+  const {setSearchQuery} = React.useContext(SelectPanelContext)
+
+  // @ts-ignore todo
+  const internalOnChange = event => {
+    // If props.onChange is given, the application controls search,
+    // otherwise the component does
+    if (typeof props.onChange === 'function') props(props.onChange)
+    // @ts-ignore todo
+    else setSearchQuery(event.target.value)
+  }
```

Comment:
Could the state for the controlled/uncontrolled search input live local in this component or are we planning that the `searchQuery` state value and setter on context will be used by other components?